### ISSUE TITLE
Wazuh Dashboard cannot connect Wazuh API after request timeout

### ISSUE THREAD

**Issue by jnasselle** [wazuh/wazuh#21422]
|Wazuh version|Component|Install type|Install method|Platform|
|---|---|---|---|---|
| v4.8.0-alpha2 | Wazuh API| Manager | OVA | Amazon Linux 2 |

## Description

During https://github.com/wazuh/wazuh/issues/21367, it was found that Wazuh Dashboard is not capable of retrieving (and therefore visualizing) large files (in this case a 1GB text file) as could be expected. This will be treated separately on another issue.

![image](https://github.com/wazuh/wazuh/assets/1791430/c82597f2-64eb-46b5-ad62-9dec4828e367)

After trying to visualize the file, the Dashboard jumps to the mail load page showing the beacon about the Timeout, but the Wazuh API check fails repeatedly. After some (generous) random time, retry works and everything is working as expected.

![image](https://github.com/wazuh/wazuh/assets/1791430/0331cd1f-1a6d-40b5-910a-644500933c84)

## Steps to reproduce

1. Create a big file into any Wazuh agent group (can be the default one)
```
yes This is a 1G Text File | head -c 1G > /var/ossec/etc/shared/deafult/group1.txt
```
2.  Using Wazuh Dashboard navigate to the group files and click the "eye" of group1.txt
![image](https://github.com/wazuh/wazuh/assets/1791430/922dfd87-e28d-49af-8d49-3d0e68e29b00)

## Extra information
api.log

```
2024/01/12 19:42:55 INFO: wazuh-wui 127.0.0.1 "GET /groups/group1/files" with parameters {"offset": "0", "limit": "15", "sort": "+filename"} and body {} done in 5.043s: 200

2024/01/12 19:43:08 ERROR: Timeout executing API request
2024/01/12 19:43:08 INFO: wazuh-wui 127.0.0.1 "GET /groups/group1/files/group1.txt/xml" with parameters {} and body {} done in 10.067s: 500
2024/01/12 19:43:09 INFO: wazuh-wui 127.0.0.1 "POST /security/user/authenticate" with parameters {} and body {} done in 0.477s: 200
2024/01/12 19:43:19 ERROR: Timeout executing API request
2024/01/12 19:43:19 INFO: wazuh-wui 127.0.0.1 "GET /manager/info" with parameters {} and body {} done in 10.083s: 500
2024/01/12 19:43:30 INFO: wazuh-wui 127.0.0.1 "POST /security/user/authenticate" with parameters {} and body {} done in 1.055s: 200
2024/01/12 19:43:30 ERROR: Timeout executing API request
2024/01/12 19:43:30 INFO: wazuh-wui 127.0.0.1 "GET /manager/info" with parameters {} and body {} done in 10.647s: 500
2024/01/12 19:43:30 INFO: wazuh-wui 127.0.0.1 "POST /security/user/authenticate" with parameters {} and body {} done in 0.675s: 200
2024/01/12 19:43:40 ERROR: Timeout executing API request
2024/01/12 19:43:40 INFO: wazuh-wui 127.0.0.1 "GET /manager/info" with parameters {} and body {} done in 10.116s: 500
2024/01/12 19:43:40 ERROR: Timeout executing API request
2024/01/12 19:43:40 INFO: wazuh-wui 127.0.0.1 "GET /manager/info" with parameters {} and body {} done in 10.061s: 500
2024/01/12 19:43:50 INFO: wazuh-wui 127.0.0.1 "POST /security/user/authenticate" with parameters {} and body {} done in 0.999s: 200
2024/01/12 19:43:50 INFO: wazuh-wui 127.0.0.1 "POST /security/user/authenticate" with parameters {} and body {} done in 0.721s: 200
2024/01/12 19:43:50 ERROR: Timeout executing API request
2024/01/12 19:43:50 INFO: wazuh-wui 127.0.0.1 "GET /manager/info" with parameters {} and body {} done in 10.034s: 500
2024/01/12 19:43:50 ERROR: Timeout executing API request
2024/01/12 19:43:50 INFO: wazuh-wui 127.0.0.1 "GET /manager/info" with parameters {} and body {} done in 10.032s: 500
2024/01/12 19:44:00 ERROR: Timeout executing API request
2024/01/12 19:44:00 INFO: wazuh-wui 127.0.0.1 "GET /manager/info" with parameters {} and body {} done in 10.099s: 500
2024/01/12 19:44:00 ERROR: Timeout executing API request
2024/01/12 19:44:00 INFO: wazuh-wui 127.0.0.1 "GET /manager/info" with parameters {} and body {} done in 10.036s: 500
2024/01/12 19:44:10 ERROR: Timeout executing API request
2024/01/12 19:44:10 INFO: wazuh-wui 127.0.0.1 "GET /manager/info" with parameters {} and body {} done in 10.048s: 500
2024/01/12 19:45:10 ERROR: Timeout executing API request
2024/01/12 19:45:10 INFO: wazuh-wui 127.0.0.1 "GET /manager/stats/analysisd" with parameters {} and body {} done in 10.046s: 500
2024/01/12 19:45:10 ERROR: Timeout executing API request
2024/01/12 19:45:10 INFO: wazuh-wui 127.0.0.1 "GET /cluster/status" with parameters {} and body {} done in 10.027s: 500
2024/01/12 19:45:10 ERROR: Timeout executing API request
2024/01/12 19:45:10 INFO: wazuh-wui 127.0.0.1 "GET /manager/stats/remoted" with parameters {} and body {} done in 10.054s: 500
2024/01/12 19:50:01 INFO: wazuh-wui 127.0.0.1 "GET /manager/stats/analysisd" with parameters {} and body {} done in 0.036s: 200
2024/01/12 19:55:01 INFO: wazuh-wui 127.0.0.1 "GET /manager/stats/analysisd" with parameters {} and body {} done in 0.040s: 200
2024/01/12 20:00:00 INFO: unknown_user 127.0.0.1 "GET /cluster/status" with parameters {} and body {} done in 0.039s: 401
2024/01/12 20:00:00 INFO: unknown_user 127.0.0.1 "GET /manager/stats/remoted" with parameters {} and body {} done in 0.019s: 401
2024/01/12 20:00:00 INFO: unknown_user 127.0.0.1 "GET /manager/stats/analysisd" with parameters {} and body {} done in 0.024s: 401
2024/01/12 20:00:01 INFO: wazuh-wui 127.0.0.1 "POST /security/user/authenticate" with parameters {} and body {} done in 0.804s: 200
2024/01/12 20:00:02 INFO: wazuh-wui 127.0.0.1 "POST /security/user/authenticate" with parameters {} and body {} done in 0.938s: 200
2024/01/12 20:00:02 INFO: wazuh-wui 127.0.0.1 "POST /security/user/authenticate" with parameters {} and body {} done in 0.556s: 200
2024/01/12 20:00:02 INFO: wazuh-wui 127.0.0.1 "GET /cluster/status" with parameters {} and body {} done in 0.217s: 200
2024/01/12 20:00:02 INFO: wazuh-wui 127.0.0.1 "GET /manager/stats/remoted" with parameters {} and body {} done in 0.172s: 200
2024/01/12 20:00:02 INFO: wazuh-wui 127.0.0.1 "GET /manager/stats/analysisd" with parameters {} and body {} done in 0.030s: 200
2024/01/12 20:00:02 INFO: wazuh-wui 127.0.0.1 "GET /agents" with parameters {"offset": "0", "limit": "1", "q": "id!=000"} and body {} done in 0.037s: 200
2024/01/12 20:00:02 INFO: wazuh-wui 127.0.0.1 "GET /agents" with parameters {"offset": "0", "limit": "500", "q": "id!=000"} and body {} done in 0.033s: 200
2024/01/12 20:05:00 INFO: wazuh-wui 127.0.0.1 "GET /manager/stats/analysisd" with parameters {} and body {} done in 0.101s: 200
2024/01/12 20:05:00 INFO: wazuh-wui 127.0.0.1 "GET /manager/stats/remoted" with parameters {} and body {} done in 0.047s: 200

```

**Comment by Selutario** [comment]
This is a known problem in the API. Two different issues come into play here:

1. The API executes the logic behind each endpoint sequentially, in a single process dedicated exclusively to this. If two (or more) requests are made in a row, the second will not start processing until the first is completed.
2. The API timeout (default, 10s) tells the client that the request was not executed on time. However, the logic behind the endpoint continues to execute until completed, regardless of whether the API has already responded with a timeout.
    
    In other words, a timeout message may appear after 10s. Still, if the function that is reading and transforming the large file to an XML takes (for example) 50s to finish, the process where it is being executed will remain busy during those 50s. It will not be able to respond to any other API requests in the API, all of them would be queued.

Therefore, what happens is that the process executing the endpoint logic is busy reading the file, and all new requests are queued up also giving a timeout until the file reading is completed (even though its result will no longer be used).

None of these problems have a simple solution. For example, canceling the task in progress could leave some communication with other services incomplete, so it requires careful study. We intend to fix the problem of sequential execution through a refactor for the correct use of asynchrony in the framework and replace the API engine, which would allow us to use multiple parallel API processes, each with its own asynchronous event loop.

We want to do this in two iterations, the first will consist of replacing the API engine and the issue to do so has already been created:
- https://github.com/wazuh/internal-devel-requests/issues/612

The second iteration would be fixing any part of the framework that may be problematic if run in parallel, refactoring all IO-blocking functions to make them asynchronous, and increasing the number of API worker processes.

**Comment by Selutario** [comment]
To be fixed in https://github.com/wazuh/internal-devel-requests/issues/612 and another future issue.
